src/kernel/sporadic.c

```diff
@@ -215,7 +215,18 @@ void refill_update(sched_context_t *sc, ticks_t new_period, ticks_t new_budget,
 static inline void schedule_used(sched_context_t *sc, refill_t new)
 {
     if (unlikely(refill_tail(sc)->rTime + refill_tail(sc)->rAmount >= new.rTime)) {
-        /* Merge overlapping refill */
+        /* Merge overlapping or adjacent refill.
+         *
+         * refill_update can produce a tail refill that will overlap
+         * with new refills when time is charged to the head refill.
+         *
+         * Preemption will cause the head refill to be partially
+         * charged. When the head refill is again later charged the
+         * additionally charged amount will be added where the new
+         * refill ended such that they are merged here. This ensures
+         * that (beyond a refill being split as it is charged
+         * incrementally) a refill split is only caused by a thread
+         * blocking. */
```

Comment:
So if a thread is preempted and later resumed, then any further budget it consumes until it blocks or runs out of budget will now be replenished 1 period in the future from when the thread was originally preempted?  This means that the sliding window bandwidth constraint is no longer enforced by the kernel too?

Reply:
Correct. The 'sliding window constraint' is incompatible with sporadic server.

Reply:
With the merging of neighboring replenishments, if I understand correctly, the replenishment effectively happens one period after a thread *resumed* (not when it was first preempted after resuming), no matter how many times it is preempted in the meantime.  Which exactly matches  sporadic server textbook behavior. Only instants when a thread blocks or runs out of budget (and hence kind of involuntarily blocks) are relevant for postponing the replenishment time.

Reply:
It's been a while since I've looked at the theory here. What is the upper bound on execution we end up with after this change? Will threads be able to cooperate to violate their temporal bounds?

Reply:
> With the merging of neighboring replenishments, if I understand correctly, the replenishment effectively happens one period after a thread resumed (not when it was first preempted after resuming), no matter how many times it is preempted in the meantime.

Right, thanks. And by merging these neighboring replenishments a preempt-able thread will no longer require storage for every potential preemption in its scheduling context object (but it will still need enough additional budget for each potential preemption for one of the kernel context switches).

What is now supposed to happen if a thread that sometimes overruns its budget is preempted for more than its period? If it eventually runs and it has been more than 1 period after it was resumed for how long will it be allowed to execute before it generates a timeout fault?

Reply:
> but it will still need enough additional budget for each potential preemption for one of the kernel context switches

That means imperfect temporal isolation. To understand how much budget a thread needs in this case, just knowing its per-activation WCET is no longer sufficient; instead, one needs to bound the maximum number of preemptions per reservation period. In turn, this requires complete knowledge of (and *trust* in) the activation behavior of all threads that could possibly preempt the thread under analysis. I think this is undesirable. 

What I've been arguing for (in emails with @xurtis some month ago) is that the cost of preemption should be charged exclusively to the *preempting* thread, so that a thread's budget needs are independent of the number of times that it is preempted (modulo cache effects such as loss of cache affinity, TLB pollution, etc.). This would also contribute "novelty" w.r.t. a potential publication on this work.

Reply:
> What is the upper bound on execution we end up with after this change?

Hi Anna, sorry, I don't quite get the question. Could you please elaborate exactly what you seek to bound? 

> Will threads be able to cooperate to violate their temporal bounds?

No. A correct sporadic server implementation does not require or assume threads to behave nicely in any way. It ensures that the maximum scheduling interference imposed by a thread on lower-priority threads never exceeds that of a Liu & Layland (or sporadic) task with equivalent parameters.

Reply:
> What is now supposed to happen if a thread that sometimes overruns its budget is preempted for more than its period?

I'm not sure what you are trying to describe here @kent-mcleod, if a thread is preempted for more than a period then whatever was able to preempted it has too high a budget and the system shouldn't have been admitted in the first place.

Reply:
> What is now supposed to happen if a thread that sometimes overruns its budget is preempted for more than its period? If it eventually runs and it has been more than 1 period after it was resumed for how long will it be allowed to execute before it generates a timeout fault?

That's a very good point actually. Normally it is assumed that admission control ensures that all servers will be able to consume their complete budget by the end of each replenishment period (i.e., the set of servers is schedulable). Then it's simple: the scenario you describe never happens. 

Now what happens if the set of servers is *not* schedulable? If the system is overloaded (sum of reservation utilization exceeds 1), then clearly the lowest-priority servers will not be able to keep up.

If the system is *not* schedulable but not overloaded, then it becomes more interesting. For instance, suppose you have a server with budget 100, period 1000 at the highest priority, and server with budget 1, period 10 at a lower priority. Total utilization is only 20%, so the system can always catch up, but clearly the lower-priority reservation can be starved for up to 10 replenishment periods. In this case, to avoid under-supplying the longterm utilization of the lower-priority server, it's important for the starving server to *accumulate* replenishments: when it finally gets a chance to run, it needs to make up for all the replenishment periods that it "missed". 

Is this supported by the current implementation? Otherwise, is it guaranteed that only schedulable sets of sporadic servers are admitted?

Reply:
Your second answer answers the first one as well. Thanks! I'm very happy you're helping the team out on this one. 

I'm also pretty excited to see the work that will charge the preempting thread for the cost of the preemption.

Reply:
> Is this supported by the current implementation? Otherwise, is it guaranteed that only schedulable sets of sporadic servers are admitted?

The intent of the design is that in the case that the sporadic servers are not schedulable, no time is guaranteed to the lower priority threads. It's up to user-level to do admission testing. The kernel makes sure that no-one exceeds their bandwidth, but provides no lower bound on execution time provided to any thread *unless* user level has done the right admission testing. 

> how long will it be allowed to execute before it generates a timeout fault?

1 * budget 

We could look at adding more options for different replenishment behaviour, however I'd need to see some compelling use cases. Most of the time these low prio threads will not be bounded in execution time at all (round robin).

Reply:
> Is this supported by the current implementation? Otherwise, is it guaranteed that only schedulable sets of sporadic servers are admitted?

If the system is correctly admitted above a certain priority level then this will not occur for tasks above that level. Below that level, the implementation will now actually provide tasks with all replenishments that were 'missed' if they were preempted such that they could not execute for some periods.

> 1 * budget (at least, that is what I intended. after these changes I'm not sure, and I would like to understand if this has changed why that's ok)

This is still the case.The updated implementaiton will

Reply:
>  Will threads be able to cooperate to violate their temporal bounds?

Currently, when a task donates its SC to a higher-priority task, the entire budget is transferred. This means that a passive server could execute continuously for the largest of its client's budgets. This makes the bound on priority inversion the largest budget of all the clients of all passive servers at an equal or higher priority than a given task.

I believe this to be untenable and will also be working on a way to ensure that a more configurable and sane bound can be placed on the execution of passive servers to bound this worst-case priority inversion.

Reply:
> If the system is correctly admitted above a certain priority level then this will not occur for tasks above that level. Below that level, the implementation will now actually provide tasks with all replenishments that were 'missed' if they were preempted such that they could not execute for some periods.

@xurtis this is changing my design and allowing tasks to violate their bandwidth constraints. What is the reasoning behind this change? The kernel design was intentionally supposed to provide an upper bound based on the SC regardless of the schedulability of the tasks.

Reply:
> I believe this to be untenable and will also be working on a way to ensure that a more configurable and sane bound can be placed on the execution of passive servers to bound this worst-case priority inversion.

Could you provide an argument for this before a design?

Reply:
> If the system is _not_ schedulable but also not overloaded, then it becomes more interesting. For instance, suppose you have a server with budget 100, period 1000 at the highest priority, and a server with budget 1, period 10 at a lower priority. Total utilization is only 20%, so the system can always catch up, but clearly the lower-priority reservation can be starved for up to 10 replenishment periods. In this case, to avoid under-supplying the longterm utilization of the lower-priority server, it's important for the starving server to _accumulate_ replenishments: when it finally gets a chance to run, it needs to make up for all the replenishment periods that it "missed".
> 
> Is this supported by the current implementation? Otherwise, is it guaranteed that only schedulable sets of sporadic servers are admitted?

I believe we are assuming RMPA, so this case should not happen

Reply:
@gernotheiser the kernel has no admission test. We do not assume *anything* about threads behaviour or configuration. The intent of the MCS design was that no matter what, you get an upper bound on execution provided by the scheduling context. What is RMPA?

Reply:
I'm aware of that, but it's not the kernel's job to fix poor design.

RMPA: rate-monotonic priority assignment, which @brandenburg's example violates.

Reply:
We don't assume rate monotonic as criticality and rate are often not the same. 

I completely disagree that the kernel guarantees for low prios should fall apart if you don't use RMPA, unless there's a strong motivation for that, but I do not see it. We're throwing away the ability to place an upper bound on threads execution no matter what priority they are set at because ... ?

Reply:
> this is changing my design and allowing tasks to violate their bandwidth constraints. What is the reasoning behind this change?

This is providing bandwidth constraints for any rate-monotonic priority-assigned task set that is admissable. The kernel guarantees that if the admission test has been done correctly than bandwidth will still be constrained. It's only in over-provisioned task sets where we lose this and only at or below the priority where the tasks at or above that priority aren't an admissable set.

The reason for this change is that it is what results from the use of sporadic server or anything that adheres to classic real-time scheduling theory. We could possibly re-introduce an upper bound on these tasks, but with added complexity in the kernel; I was of the impression this was not an explicit goal of the design.

> Could you provide an argument for this before a design? The client has to trust a server that it's blocking on to obey it's WCET, which is the actual bound.

This is an issue with _any passive server at a higher priority_, not just those to which the task is a client. To guarantee temporal isolation in the presence of untrusted, higher-priority passive servers we must enforce a bound on priority inversion, which means a bound on the execution that server can do on behalf of any lower-priority clients.

Reply:
> We could possibly re-introduce an upper bound on these tasks, but with added complexity in the kernel

I'm also uncertain we could do this correctly without adding loops to the kernel that are effectively unbounded.

Reply:
@xurtis handling overprovisioning was specifically part of the goal. Scheduling contexts were supposed to create an upper bound on your execution, no matter what, as it's a capability to execution time on the CPU. I did intentionally tweak the sporadic server model for this. If that reduces the practicality of the system due to practical utilisation we can look at relaxing it, but this needs to be very explicit because you are changing the model. 
 
Can you give a use case for an untrusted, high-priority server? Is this because you are now assuming RMPA (which is another change to the model that needs to be explicit. Seems to me you are going to *want* a scheduling test in the kernel if that is what it takes to be correct  - e.g there is now only one test that is sufficient and the SC's won't work otherwise.)

Reply:
Is it more important to guarantee constant bandwidth on all tasks or to be able to use the SCs for ensuring sufficient temporal isolation that we can perform some kind of scheduling analysis in task sets with untrusted tasks?

I can't see the former being valuable at all if it means we can't have the latter and that we must provision at least twice as much bandwidth as we ever expect them to use in a way that either halves maximum utilisation on does not actually temporally isolate tasks.

Reply:
>  if it means we can't have the latter and that we must provision at least twice as much bandwidth as we ever expect them to use in a way that either halves maximum utilisation on does not actually temporally isolate tasks.

Can you please explain what this is a consequence of?

> guarantee constant bandwidth on all tasks

not a constant bandwidth, but an upper bound.

Reply:
> > What is now supposed to happen if a thread that sometimes overruns its budget is preempted for more than its period? If it eventually runs and it has been more than 1 period after it was resumed for how long will it be allowed to execute before it generates a timeout fault?
> 
> That's a very good point actually. Normally it is assumed that admission control ensures that all servers will be able to consume their complete budget by the end of each replenishment period (i.e., the set of servers is schedulable). Then it's simple: the scenario you describe never happens.
> 
> Now what happens if the set of servers is _not_ schedulable? If the system is overloaded (sum of reservation utilization exceeds 1), then clearly the lowest-priority servers will not be able to keep up.
> 
> If the system is _not_ schedulable but also not overloaded, then it becomes more interesting. For instance, suppose you have a server with budget 100, period 1000 at the highest priority, and a server with budget 1, period 10 at a lower priority. Total utilization is only 20%, so the system can always catch up, but clearly the lower-priority reservation can be starved for up to 10 replenishment periods. In this case, to avoid under-supplying the longterm utilization of the lower-priority server, it's important for the starving server to _accumulate_ replenishments: when it finally gets a chance to run, it needs to make up for all the replenishment periods that it "missed".
> 
> Is this supported by the current implementation? Otherwise, is it guaranteed that only schedulable sets of sporadic servers are admitted?

It seems that in the scenario described (according to my interpretation of the changes in this PR), after the starving server starts executing after preemption by the 100 budget server, it will be able to consume the remaining budget from the period that it was preempted in, and then will be able to be released immediately due to an available refill.  When being _released_, `refill_unblock_check` is called by the kernel and the head refill will be delayed to the current value of `ksCurTime`.  It then follows that when the task consumes it's budget again it will not be able to continue executing until the period completes.  Based on this, it appears that the starving server will not _accumulate_ replenishments so that it can catch up. Is it supposed to?

Reply:
> If the system is _not_ schedulable but also not overloaded, then it becomes more interesting. For instance, suppose you have a server with budget 100, period 1000 at the highest priority, and a server with budget 1, period 10 at a lower priority. Total utilization is only 20%, so the system can always catch up, but clearly the lower-priority reservation can be starved for up to 10 replenishment periods. In this case, to avoid under-supplying the longterm utilization of the lower-priority server, it's important for the starving server to _accumulate_ replenishments: when it finally gets a chance to run, it needs to make up for all the replenishment periods that it "missed".

Why does it make sense with "accumulate" replenishments at all? If this is a classic "periodic task" with WCET 1 and period 10 then "catching up" doesn't really make sense does it? If it needed to service an interrupt once every 10ms, you generally can't expect to be able to just service 10 interrupts all at once and "catch up". If that was feasible, shouldn't it just have been assigned budget 10, period 100 in the first place?

Reply:
Here’s an interpretation of “periodic task” where it the catching-up behavior happens naturally: https://sigbed.org/2020/09/05/liu-and-layland-and-linux-a-blueprint-for-proper-real-time-tasks/

Reply:
>  It then follows that when the task consumes it's budget again it will not be able to continue executing until the period completes.

A task is only _released_ if it unblocks or transitions from having no available replenishments to any at all. With this PR, the thread would not do either of these if it were to completely consume the budget from replenishments multiple periods in the past. When such a replenishment is consumed, the new replenishment would be placed one period later which would still be in the past and thus still usable immediately.

Reply:
> Why does it make sense with "accumulate" replenishments at all? If this is a classic "periodic task" with WCET 1 and period 10 then "catching up" doesn't really make sense does it? If it needed to service an interrupt once every 10ms, you generally can't expect to be able to just service 10 interrupts all at once and "catch up". If that was feasible, shouldn't it just have been assigned budget 10, period 100 in the first place?

Does it make sense to over-commit the system such that this task couldn't service all of those interrupts?

Additionally, if there is an associated amount of work with each interrupt (e.g. processing an individual network packet) and that many interrupts had occurred, it would make sense to provide time to process all of the packets that had arrived in that time rather than only a single packet.

Reply:
Let me elaborate a bit more. First, I agree with Anna that it’s undesirable to bake assumptions about rate-monotonic priorities into the kernel. There are (famous) cases where RM is a good (or even optimal) approach, but there also valid use case where it is not. Even if my earlier example is contrived, it’s not unreasonable to expect that for some servers *bounded tardiness* is sufficient, a “soft” notion of timeliness, which my example satisfies.

Let’s distinguish between short-term and long-term guarantees. 

The classical notion of hard schedulability is a short-term guarantee: a  set of sporadic servers is schedulable if every server instance is guaranteed to be able to consume its entire budget in each server period. From this, one implicitly gets a long-term guarantee “for free” — in the long run, the  a server is indeed guaranteed a fraction of processor service matching its utilization (budget divided by period). Since the latter long-term guarantee is implied by the commonly studied short-term guarantee, the long-term guarantee is not often discussed. 

Now what if a set of reservations is not (hard) schedulable? As a long as the core is not overloaded (total utilization exceeding 1), one still gets a *bounded tardiness* guarantee: every server instance is able to consume its entire budget within a fixed time window  of *constant*, *bounded* length  starting with the activation of the instance (but not necessarily always *during* its period). This is equivalent to saying that any server will *lag* behind its promised long-term allocation by only a bounded, constant amount.

For “soft” or low-criticality workloads, such a bounded-tardiness guarantee (“in the long run, it all averages out”) can often be perfectly acceptable. What’s not acceptable is to accumulate *unbounded* tardiness, i.e., situations where the workload never catches up because it is systematically under-provisioned, which leads to ever growing input queues (or eventually dropped activations / skipped data). 

*Any* plain old fixed-priority scheduler (without reservations, e.g., SCHED_FIFO in Linux) provides bounded tardiness on  a uniprocessor as long as the processor is not overloaded: if a task misses some deadlines and builds up a backlog of activations (as in the linked Linux example above), it will remain continuously ready for execution and eventually catch up. In case of periodic tasks with a synchronous release, at each hyperperiod boundary no task will have any backlog even if the task set is not schedulable. 

By not accumulating reservation budgets during times of starvation, seL4 destroys this very useful property. Essentially, you get an all-or-nothing proposition: either a server is hard-schedulable, in which case one gets the short-term guarantee, or it is not, in which case one gets *no* guarantee at all, not even one of bounded lag / bounded tardiness / a lower bound on guaranteed longterm utilization (which all imply each other, i.e., different ways to state the same guarantee). 

To me, this seems like an unnatural design choice. FP/sporadic servers inherently give you bounded tardiness (on uniprocessors and under partitioned scheduling), which is a very useful property, so why intentionally throw it away?

Reply:
> Can you please explain what this is a consequence of?

I've made a [spreadsheet](https://docs.google.com/spreadsheets/d/183E_kGfN3ooVoW-8vciuosLNfgJ6cPagiMBWR5_ZWR8/edit?usp=sharing) describing a particular scenario where a task set that _should_ be schedulable with standard _rate-monotonic fixed-priority_ (RMFP) scheduling cannot be scheduled using the fixed-bandwidth constraints provided in the existing implementation of SCs.

The first case (normal periodic execution) describes an aperiodic task with a minimum inter-arrival time of 3 and WCET 1 and a lower-priority periodic task of period 6 and WCET 1. This shows that this set of periodic tasks would expect the lower priority task to require 2x its WCET in bandwidth under the _sliding window constraint_ in a case where the higher priority task has a delayed aperiodic activation.

The second case (constant bandwidth enforced) demonstrates that when we apply the constant bandwidth constraints, delayed aperiodic activations in the higher priority task can cause the lower-priority task to accumulate delay until it completely misses a deadline.

The third case (lower-priority tasks w/ increased bandwidth 1) shows the increase of the period 6 task's budget to allow it to use up to twice the bandwidth and the introduction of a new periodic task with a period of 8 and a WCET of 1. This shows that the period 8 task would need 2x its WCET in bandwidth under the sliding window constraint for the same reasons the period 6 task would.

The fourth case (constant bandwidth exhausted 1) if the period-6 task were to always exhaust available budget as soon as possible and the period-8 task were only provided budget = WCET. The period-8 task can miss deadlines.

The fifth case (lower-priority tasks w/ increased bandwidth 2) increases the budget of the period-8 task and adds a periodic task of period 16 with a WCET of 2. It shows that when all tasks are well behaved, the period-16 task would need up to 1.5x its WCET in bandwidth to work under the 'sliding window constraint'.

At this point, we have a set of tasks scheduled with rate-monotonic fixed-priority that, were they constrained to execute bounded by aperiodic task execution (or sporadic server) would be guaranteed to always meet deadlines as the set satisfies the basic admission test for RMFP (`1/3 + 1/6 + 1/8 + 2/16 = 0.75 < 4*(2^(1/4) - 1)`). However, to ensure they can still execute in this manner in the presence of the sliding window constraint, for all but the highest priority task, we have to provide a greater budget than simply the WCET/minimum inter-arrival time and have over-committed the CPU time (`1/3 + 2/6 + 2/8 + 3/16 > 1.104`).

In fact, the total bandwidth provided to the top three tasks alone ('1/3 + 2/6 + 2/8 > 0.916`) doesn't leave sufficient time for even the WCET/inter-arrival time of the lowest priority task (`2/16 = 0.125`). The result of this is seen in the final case. Despite this task set being schedulable by the RMFP admission test, when the higher-priority tasks consume all time made available to them as soon as possible, the lowest-priority task is no longer guaranteed to meet deadlines.

This example demonstrates that when the sliding-window constrain is applied, one must sacrifice either utilisation (and determine some new appropriate admissions test and loosing up to 50% of utilization to time budgeted for the sliding window constraint that aught never to be used) or temporal isolation (and trust that higher-priority tasks won't exceed their WCET even when they have the budget to do so) in order for this to still be sufficient in any hard real-time system.

I made have made some mistakes in this example so I'd appreciate if I can get some confirmation there, but this is why I understand the sliding-window constraint to be incompatible with hard real-time task sets and a misfeature in a mixed-criticality real-time scheduler.

Reply:
That isn't to say that I think one way or the other on whether a low-priority task should reclaim all lost time in an over-committed system or still be constrained to just lose any time from periods where it was unable to exhaust it before the period expired. I can think of good reasons why either may be desirable in a particular system. The result of this PR would be that a task would accumulate lost time, but that is by happenstance and not a conscious decision.

I suspect that properly expiring budget with periods exactly would be complicated although approximating that behavior could be done with little added complexity. I don't know whether such an approximation would be sufficient for the cases where this would occur though.

Reply:
Thanks for the explanations! It has made it easier to think through the consequences of the existing design and the proposed changes.  This is where I'm at now:

If you let the old SC behavior be A and the new SC behavior be B, then previously the kernel was (A and (not B)).  This PR changes the kernel to be ((Not A) and B).  I think there's a bit of debate that's combining the following issues:
- Whether B is behavior we want to have
- Whether A is behavior that we still want
- Whether A and B are mutually exclusive.
- If A and B are mutually exclusive, whether we want B more than A.

**Whether B is behavior we want to have**
It's clear to me that the ability to catch-up is a useful and often necessary feature to have and that having this as a core behavior of scheduling contexts removes a significant barrier to building useful systems with analyzable timing properties.  I don't completely agree with Curtis' statement that certain real-time systems cannot be implemented with the existing SC mechanisms because it assumes a system where you _only_ have configured scheduling-context (SC) mechanisms and don't use the timeout fault handlers and a user level scheduler which were also provided to implement timing guarantees. That said, in typical high-assurance systems, where you want to have as minimal dynamism as possible, being able to have a simple system initializer set up and create all of the caps implementing your desired timing policies (with offline pre-computed admission checks) before self-terminating and not needing timeout fault handlers or a user level scheduler is a goal worth pursuing. However I can't see how this type of system would be able to implement over-commit + mixed criticality without fault handlers and user scheduling. 

Additionally, in its current form, it seems hard for the kernel to enforce that an entity can't abuse the catch-up semantics for violating time protection.  The kernel now distinguishes whether a thread is blocked or preempted and allows it to accumulate budget across periods if it is preempted, but not if it is blocked.  But being in a preempted state was never previously something the kernel tried to restrict and so it's very easy for a thread to put its SC onto a lower priority thread that it knows will be preempted for large amounts of time (lower priority than any round robin thread for example) and steadily accumulate an unbounded amount of budget. For example, instead of only blocking near the end of its budget, a thread can also once blocked, rely on a secondary thread to move the SC to another thread of a low priority.  There's currently no way to restrict creating a thread with the lowest priority level (other than by limiting access to untyped) or by setting an existing thread to the lowest priority level (other than not having a cap to that thread). Restricting that would also restrict a wide amount of unrelated system behavior.  Then there's no way to restrict binding an SC to this arbitrary thread if one has the cap to the SC and the thread.  Alternatively, it's possible to use an endpoint to donate an SC to a lower priority thread without caps to the SC or the target thread. Then once enough time has passed, the lower-criticality, but higher priority thread can execute for a much longer time continuously than was specified.
I can't think of an easy way to mitigate this. 

**Whether A is behavior that we still want**
I think in dynamic open systems, where timeout fault handlers and a user-level scheduler implement the scheduling policy and use the scheduling contexts to implement an upper bound on the execution time of a task for the kernel to enforce, having a strict upper bound on the short-term bandwidth is still something that I think many systems would want to be able to do. In addition, this behavior has been the only bandwidth constraining policy available for the last couple years and removing it will potentially cause discomfort to people who have architect-ed systems around it.

**Whether A and B are mutually exclusive**
I think it is possible to implement either in the other with the use of timeout faults and dynamic user level scheduling, but I don't think it's possible to implement either in the other within a static system.  However, I don't think its a foregone conclusion that the kernel can only have a single type of replenishment policy that it will enforce.
It is also possible to add a new replenishment policy as a new scheduling context configuration parameter.  We already have 2 types, and it doesn't seem prohibitive to add a third. Then a system designer can choose how which replenishment policy for the kernel to enforce for each scheduling context.  Given that verification costs are pretty high, it's likely that we can only verify one type in the near future right (and B is already in the middle of being verified right?)

**If A and B are mutually exclusive, whether we want B more than A**
I think some people would find it controversial to not pick B and also I think other people would find it controversial to not pick A.

Reply:
@brandenburg @kent-mcleod @xurtis thanks for the really clear discussion on this and time and effort! My 2 cents: I definitely toyed around with the idea of selecting different replenishment policies based on configuration in the SC. This is effectively what we do already with round robin vs sporadic, and we could add another which is bounded.

Exploring how to compromise sporadic server guarantees by using cooperating threads and preemption would be a cool project!

Reply:
> I don't completely agree with Curtis' statement that certain real-time systems cannot be implemented with the existing SC mechanisms because it assumes a system where you only have configured scheduling-context (SC) mechanisms and don't use the timeout fault handlers and a user level scheduler which were also provided to implement timing guarantees.

If you're relying on a user-level scheduler to enforce these guarantees, what point is there enforcing them in the kernel at all?

> I can't think of an easy way to mitigate this.

It's not clear to me why we are pursuing a definition of 'temporal isolation' that ensures threads don't exceed some specified bandwidth anyway, if you've over-comitted the system you won't get much of a useful guarantee from this and if you're building a real-time system this guarantee isn't sufficient or useful either.

> I think it is possible to implement either in the other.

As I was trying to show in my example, the existence of the bandwidth constraint either means you can't implement your hard real-time system or the kernel's concept of 'temporal isolation' is in no way useful to you implementing it, you end up having to over-provision budget and either trust all of your tasks or enforce the actual temporal guarantees you need at user level; if you ever hit the case where you violate the kernel's concept of temporal isolation you're system is already well past faulted.

> and B is already in the middle of being verified right

B is what is being verified as that was my best understanding of what we were actually trying to achieve.

> I think some people would find it controversial to not pick B

A large portion of the facilities introduced in MCS seem basically useless for real-time systems if the temporal isolation guarantees provided by the kernel are so much weaker than what is required by any real-time scheduling policy. Not to mention, as soon as you need to enforce any hard real-time guarantees over IPC boundaries, you either lose the ability to enforce them or need to interpose every IPC with your user-level scheduler.

Reply:
> It's not clear to me why we are pursuing a definition of 'temporal isolation' that ensures threads don't exceed some specified bandwidth anyway, if you've over-comitted the system you won't get much of a useful guarantee from this and if you're building a real-time system this guarantee isn't sufficient or useful either.

Because we are building for open-systems, and systems of different guarantees on timing. We want to be able to say that untrusted thread behaviour will not violate temporal isolation guarantees. And you cannot trust threads to not overcommit. Otherwise you're assuming total awareness of scheduling behaviour, which is not practical in a real mixed-criticality system.

Reply:
> This is effectively what we do already with round robin vs sporadic, and we could add another which is bounded.

This sounds like it would be able to widen what we support directly in the kernel to include constant-bandwidth server-like systems that restrict lower-priority tasks to constant bandwidth but schedule higher-priority, hard-realtime tasks as they would traditionally be scheduled.

Reply:
> > I can't think of an easy way to mitigate this.
> 
> It's not clear to me why we are pursuing a definition of 'temporal isolation' that ensures threads don't exceed some specified bandwidth anyway, if you've over-comitted the system you won't get much of a useful guarantee from this and if you're building a real-time system this guarantee isn't sufficient or useful either.

What I described allows a thread to run continuously for a longer duration than its budget allows by tricking the kernel to think it has spent time being preempted when it was actually blocked.  This may not violate the average bandwidth of the SC, but it would be able to cause lower priority threads to miss their deadlines which seems like a temporal isolation violation.

Reply:
> I'm not sure what value we can get from timeout faults in a system where they would occur well after the timing guarantees being enforced by a user-level scheduler. If the kernel's concept of temporal isolation is substantially weaker than what is required by a scheduler (due to needing to provide more time to an SC to allow it to execute multiple times in a sliding window) then It's hard to leverage in implementing a user-level scheduler that needs to support hard real-time tasks.

I thought Seb's thesis showed it was possible to construct user level schedulers that support HRT tasks? https://ts.data61.csiro.au/publications/theses_public/18/Holzapfel:be.abstract

Reply:
Right, you can do that, but I'm not sure what benefit we get from the additional timing constraints in the kernel. I guess you can use it to constrain tasks that are globally untrusted to execute in a real-time fashion, but that seems more restrictive than would be desireable.

Reply:
I think we might be able to enforce a notion of temporal isolation that directly guarantees that assumptions made by EDF and RM scheduling algorithms that can have a constraint added to prevent the issues that @pingerino has raised regarding tasks co-operating to increase bandwidth.

 1. We guarantee that a job will receive execution budget after each activation and that it can only execute while it has sufficient budget
 2. (for some SCs) we guarantee that a jobs execution budget will expire one period - budget after its activation; if it is preempted sufficiently long enough, it would only be able consume budget up to the end of that period.

(1) should be sufficient for implementing EDF and RMFP (and perhaps others). If (2) is applied to an SC, that would constitute an ongoing bandwidth constraint and prevent the accumulation of any budget while a task is preempted. If (2) is not applied to an SC, that SC would be able to 'catch-up' in the manner described by @brandenburg.

I refer to 'jobs' above rather than tasks to reflect that this would be applied to individual replenishments rather than a the SC as a whole, this would be in keeping with the sporadic server model which allows a task to effectively split its time into multiple distinctly scheduled tasks with the same period and which we implement (and bound) with replenishments.

Reply:
I'm getting lost in this discussion, I think this is now touching on several different aspects, including major design choices that might be better discussed in a more structured environment. 

Anyway, let me offer my perspective and the points that confuse me:

A and B are definitely two very different interpretations. B is consistent with 20+ years of work on "temporal isolation" in the real-time community. A does not match any notion of "temporal isolation" that I'm aware of. I consider A to be analytically pretty useless (I don't see how to give strong guarantees, I think worse examples than what @xurtis illustrated are possible with analytically usable utilization approaching zero). 

I think it would be a severe mistake to choose A *instead of* B; I would not consider that to be a viable platform for real-time MCS. If there are strong reasons keep A, then the cleanest solution would indeed to just make this a completely different class of threads, rather than trying to graft A-like behavior onto a well-established, classic RT concept that simply doesn't work that way. 

> Then once enough time has passed, the lower-criticality, but higher priority thread can execute for a much longer time continuously than was specified.

This whole line of reasoning is confusing to me. I think this is part of a larger misunderstanding (likely on my behalf) — what is the contract? The guarantee that you want to accomplish? There seem to be more assumptions/requirements than I'm aware of. What do you actually mean by "temporal isolation"? We seem to be working based on differing implicit definitions of the term. 

> We want to be able to say that untrusted thread behaviour will not violate temporal isolation guarantees. 

Sporadic servers give you *exactly* that.  I feel like I'm missing something here. 

> Exploring how to compromise sporadic server guarantees by using cooperating threads and preemption would be a cool project!

I don't see how that would be possible. Can you define "cooperating" and "compromise"? What potential problem do you see? 

> What I described allows a thread to run continuously for a longer duration than its budget allows by tricking the kernel to think it has spent time being preempted when it was actually blocked. 

First, I don't see how that tricking would work. What matters is not what the thread does, but how the reservation behaves. If you move the blocked thread out of the reservation while the reservation's budget is then used to run a server the thread is blocked on, it has (or at least should have) absolutely no impact on the behavior of the reservation, and hence on the timing guarantees overall. 

Second, why limit the maximum time that a thread can continuously run when it is catching up? I see no analytical reason to do this. What do you think the kernel needs to defend against here? 

> This may not violate the average bandwidth of the SC, but it would be able to cause lower priority threads to miss their deadlines which seems like a temporal isolation violation.

This I dispute. That would mean the sporadic server algorithm is broken, which I don't think it is. Can you construct a counterexample on paper?

Reply:
> I'm getting lost in this discussion, I think this is now touching on several different aspects, including major design choices that might be better discussed in a more structured environment.

Thanks for still taking the time to consider my points. I'm still trying to understand the what the effects of allowing tasks to catch up will have on the contract that system designers have with seL4 in order to get some guarantees.  I believe that the PR introduces new requirements that system designers would have to fulfill that currently aren't met by a large range of existing seL4 system designs.

> What do you actually mean by "temporal isolation"?

Where the temporal behavior of one task cannot cause temporal faults in another, independent task.  In my specific example, a "hard" real-time task that has a job that misses its deadline would be considered a temporal fault.  If the task was operating in a configuration where all other tasks at its priority and higher were schedulable, but it missed its deadline because one of these tasks were able to change the system scheduling behavior such that the task was able to execute continuously at a higher priority for longer than the schedule allowed then this would be an example of a temporal fault being caused by another independent task and a temporal isolation violation.

> This whole line of reasoning is confusing to me. I think this is part of a larger misunderstanding (likely on my behalf) — what is the contract? The guarantee that you want to accomplish? There seem to be more assumptions/requirements than I'm aware of.

I expect that we are starting from different assumptions about the contract that system designers have with the kernel in order to get it's guarantees. In particular, controlling which priorities that a each reservation is able to run at. There exist task sets that are schedulable that can become not schedulable if the priorities of the tasks are changed right? Therefore it is expected that the system prevents tasks from executing at priorities that are different from what they are assigned in the schedule. In the case of seL4 this means that system designers have a requirement that they configure the system such that untrusted tasks don't have permissions that allow them to modify their scheduling parameters. One requirement would be that a task isn't able to execute at a priority higher than allowed and it doesn't have access to any other resources that it can coerce to execute at a higher priority using its reservation.  I believe that you now need to assume another requirement: that a task isn't able to execute at a priority lower than allowed as this effectively reassigns the priorities in a schedule and can make the system not schedulable.

The requirement to have a lower bound on a task's priority is a lot harder to achieve in seL4 than an upper bound.  The kernel requires no additional permissions to set a thread to the lowest priority as long as a capability to the thread can be accessed. The kernel requires no additional permissions to create a new thread object at the lowest priority provided that there is access to an untyped object of sufficient size. Therefore if an untrusted task can meet one of these requirements it is pretty easy for it to move it's reservation onto the thread of the lowest priority and move it back at a later point.  I think that in order to prevent this, the system designer would need to treat all code that can access either thread, untyped or SC capabilities as at least as critical as any tasks that wanted temporal isolation guarantees.

> Can you construct a counterexample on paper?

Here is an example of a system with 3 tasks.  The system is not overloaded.  The system is schedulable if the tasks are assigned priorities using RMS but it is not schedulable if the task with the highest activation rate is instead given the lowest priority.  The 3rd trace shows how the highest priority task can cause a temporal fault in the 2nd highest priority task by causing its reservation to execute at a lower priority until the system is idle, at which point the lower priority thread returns the reservation to the original thread and it can execute continuously such that the 2nd highest priority task misses several deadlines.
https://docs.google.com/spreadsheets/d/1dnGwfLMqbeYoEtATrJLhlmJyPH0J23nAQEaZcTBAJjk/edit#gid=0

> What do you think the kernel needs to defend against here?

I think the kernel needs to provide a way of limiting the minimum priority a thread can set another thread to the same way that it limits the maximum priority a thread can set another thread.  In addition, a new thread should start with an invalid priority that prevents it from being scheduled. Fixing this ambient authority is kind of unrelated to the proposed changes in the PR, but the proposed changes now allow the ambient authority to be used to "violate temporal isolation" in any case where the system designer doesn't consider preventing access to any thread, SC or Untyped capabilities by untrusted components as part of the contract with the kernel (which is a large percentage of existing projects I would claim).

Of course, I'd prefer it if I've made some sort of mistake or misunderstanding in my assumptions or reasoning for all this as I understand that a lot of work has gone into designing and implementing these changes over the last few months. Otherwise, it would be good to at least have a proposed plan for mitigating this issue before merging.

Reply:
I suspect there might be a general assumption in fixed-priority scheudles that a _task_ doesn't ever lower its priority, although priority inheritance and priority ceiling protocols can cause a task to temporarily raise its priority. I think in general for real-time you don't want to delegate the ability to control priority at all, the scheduling algorithm has strict requirements on priority assignment and can't function if not in complete control of it.

> https://docs.google.com/spreadsheets/d/1dnGwfLMqbeYoEtATrJLhlmJyPH0J23nAQEaZcTBAJjk/edit#gid=0

I'm not sure what admissions test you're using here but I don't that task seems to exceed the schedualable utilisation described by Liu & Layland (`n*(2^(1/n) - 1`) for 3 tasks.

Reply:
> I don't see how that would be possible. Can you define "cooperating" and "compromise"? What potential problem do you see?

I am thinking of a scenario in a dynamic system, where an untrusted thread, which is using a scheduling context to bound it's execution rate, can down-call to a low priority thread, accumulate budget, and then violate the temporal isolation guarantees of the system? 

I believe this is what @kent-mcleod  is describing. If that's the case, then this change means that such untrusted, high priority threads can be given no resources that would create threads without some sort of lower-limit on the priorities that an SC can execute at.

Reply:
Could this be resolved by treating a reply from a lower or equal priority thread as an activation and to just move all replenishments forward to the time of the reply when this occurs?

Reply:
Thanks for the example. It's a lot easier for me to see what you mean in a concrete trace. A couple of points. 

1. Generally speaking, under reservation-based scheduling, it doesn't make sense to talk about *thread* priorities. Reservations have priorities, the highest-priority active reservation is selected for service, and then in turn dispatches *any* of its clients. Thread priorities only come into play only if you want to enforce some order of execution among clients *of the same reservation*. Priorities of threads in different reservations are never compared and are meaningless. (For example, see the P-RES plugin in LITMUS^RT for a concrete example.)
2. Note that the sporadic server algorithm that you implemented is a *fixed*-priority algorithm. So, yes, if you start to change the priorities of the reservations at runtime, all guarantees go out the window.  (Aside: there even exists a "priority exchange server" in the literature, but it works differently from sporadic servers. IMHO sporadic servers are the better, more elegant solution.)
3. Changing the scheduling parameters of a set of servers (such as budget, replenishment period, or priority), or adding or removing reservations, is called a *mode change*. There's an entire literature on mode-change protocols that ensure that all timing guarantees are valid before and after a mode change, and crucially also in the transient phase *during* a mode change. Not all mode changes can be enacted instantly.
4. My general recommendation is to keep the mode-change issue as simple as possible by not trying to be very aggressive in enacting mode changes quickly, but fancy approaches that seek to minimize enactment delay exist, too.
5. There exist special cases where the provided guarantees are not affected (or where the effects are easy to incorporate into  a single static analysis). Examples include the *bandwidth inheritance protocol* and some other synchronization protocols that are not worth getting into at the moment. 

Bottomline: what you describe in your example may be possible behavior in seL4 (I don't know, but don't doubt it), but indicates an imperfect/incomplete/inconsistent/misunderstood/misaligned/??? integration of the sporadic server algorithm into the kernel. I remain convinced that sporadic servers can be a great fit for seL4, but the way  "schedule contexts" map to "reservations", what entities actually have priorities that the top-level scheduler cares about, and how and when it is ok to "donate" a thread's scheduling context to another thread need some more thought and discussion.